lib/Core/PTree.h

```diff
@@ -25,40 +29,49 @@ namespace klee {
   constexpr int PtrBitCount = 3;
   using PTreeNodePtr = llvm::PointerIntPair<PTreeNode *, PtrBitCount, uint8_t>;
 
+  enum class PTreeNodeMode {
+    NONE = 0,
+    NeedsUpdate = 1U,
+  };
+
   class PTreeNode {
   public:
-    PTreeNode *parent = nullptr;
-
-    PTreeNodePtr left; 
+    PTreeNode *parent {nullptr};
+    PTreeNodePtr left;
     PTreeNodePtr right;
-    ExecutionState *state = nullptr;
+    ExecutionState *state {nullptr};
+    std::uint64_t id {0};
```

Comment:
This is a veryyyyy luxurious use of fields. Isn't most of this derived from `state`? So you don't need these fields. If they are different you should put them on a separate memory allocation so they don't inflate this object too much. Some version of #1284 can still be useful that way.

Reply:
We talk about a few more MB for a huge tree vs. many allocations. `id` and `terminationMask` are needed to update nodes in the database. And `mode`+`branchType`+`asmLine` could be stored externally but are in sum smaller than a pointer. :man_shrugging:

Reply:
Doesn't it matter for caches?

Reply:
A tree is not very cache-efficient... besides the vector version in `klee-ptree`. ;)